### PR TITLE
feat(ph4): inclusion proofs

### PR DESCRIPTION
**Pull request by vladopajic**
### Checklist

- [x] I have read the [coding guide](https://github.com/ethersphere/bee/blob/master/CODING.md).
- [ ] My change requires a documentation update, and I have done it.
- [x] I have added tests to cover my changes.
- [x] I have filled out the description and linked the related issues.

### Description
This PR implements inclusion proofs for claim method of phase 4 redistribution smart contract.

Note: in code there is `// TODO_PH4` which should be handled before PR is merged to master, please go through these markers in code and handle them accordingly. 

Closes:  #3800

### CODE REVIEW

**Review by mrekucci — COMMENTED** [review]
It would also be good to include a meaningful description of all public structures/functions/methods/types...

File: pkg/storageincentives/proof.go [hunk 1]
```diff
@@ -5,11 +5,208 @@
...
+	"github.com/ethersphere/bee/pkg/postage"
+	"github.com/ethersphere/bee/pkg/storageincentives/redistribution"
+	"github.com/ethersphere/bee/pkg/storageincentives/types"
+	. "github.com/ethersphere/bee/pkg/storageincentives/types"
+	"github.com/ethersphere/bee/pkg/storer"
 	"github.com/ethersphere/bee/pkg/swarm"
 )
 
+// This file implements logic described here:
+// https://hackmd.io/C7EvRbghQpWr9UXiGOFgLQ?view
```
[the remark this answers is not in the document]
Comment (vladopajic) [reply pkg/storageincentives/proof.go:24]: It should be accessible, could you double check?

File: pkg/storageincentives/proof.go [hunk 2]
```diff
@@ -5,11 +5,208 @@
...
+// makeInclusionProofs creates transaction data for claim method.
+// In the document this logic, result data, is also called Proof of entitlement (POE).
+func makeInclusionProofs(
+	reserveSampleItems []storer.SampleItem,
+	anchor1 []byte,
+	anchor2 []byte,
+) (redistribution.ChunkInclusionProofs, error) {
+	proofs := redistribution.ChunkInclusionProofs{}
+
+	if len(reserveSampleItems) != 16 {
```
[the remark this answers is not in the document]
Comment (vladopajic) [reply pkg/storageincentives/proof.go]: fixed

File: pkg/storageincentives/proof.go [hunk 3]
```diff
@@ -5,11 +5,208 @@
...
+	wcp.A = int(new(big.Int).Mod(anchor2Big, big.NewInt(15)).Int64())
+	wcp.B = int(new(big.Int).Mod(anchor2Big, big.NewInt(14)).Int64())
+	wcp.C = 15
+
+	wc := Trio[storer.SampleItem]{} // wc = witness chunk
+	wc.A = reserveSampleItems[wcp.A]
+	wc.B = reserveSampleItems[wcp.B]
+	wc.C = reserveSampleItems[wcp.C]
+
+	wp, err := makeProofTrio( // wp := witness chunks proof
```
[the remark this answers is not in the document]
Comment (vladopajic) [reply pkg/storageincentives/proof.go:67]: These variables are used only scope local to function where they were declared, right?

File: pkg/storageincentives/proof.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ldeffenb) [reply pkg/storageincentives/proof.go:24]: I can view this document externally.

File: pkg/soc/soc.go [hunk 4]
```diff
@@ -66,13 +66,28 @@ func (s *SOC) WrappedChunk() swarm.Chunk {
...
+
+// OwnerAddress returns the ethereum address of the SOC owner.
+func (s *SOC) OwnerAddress() []byte {
+	return s.owner
+}
+
+// ID returns the SOC id.
+func (s *SOC) ID() []byte {
+	return s.id
+}
```
Comment (nugaon) [inline pkg/soc/soc.go:89]: why not public?

File: pkg/storageincentives/redistribution/inclusion_proof.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+		ProofSegments2: [][32]byte{types.ToByte32(testutil.RandBytes(t, 32))},
+		ProveSegment2:  types.ToByte32(testutil.RandBytes(t, 32)),
+		ProofSegments3: [][32]byte{types.ToByte32(testutil.RandBytes(t, 32))},
+		ChunkSpan:      1,
+		Signature:      testutil.RandBytes(t, 32),
+		ChunkAddr:      types.ToByte32(testutil.RandBytes(t, 32)),
+		PostageId:      types.ToByte32(testutil.RandBytes(t, 32)),
+		Index:          1,
+		TimeStamp:      uint64(time.Now().Unix()),
+	}
```
[the remark this answers is not in the document]
Comment (nugaon) [reply pkg/storageincentives/redistribution/inclusion_proof.go:63]: additionally, this sample test data depends on the anchors so the test definitely needs a sample generator algorithm based on those.

**Review by nugaon — CHANGES_REQUESTED** [review]

File: pkg/storageincentives/proof.go [hunk 6]
```diff
@@ -5,11 +5,234 @@
...
+	socAddr, err := socCh.Address()
+	if err != nil {
+		return nil, err
+	}
+
+	return []redistribution.SOCProof{{
+		Signer:     common.Address(socCh.OwnerAddress()),
+		Signature:  socCh.Signature(),
+		Identifier: types.ToByte32(socCh.ID()),
+		ChunkAddr:  types.ToByte32(socAddr.Bytes()),
```
[the remark this answers is not in the document]
Comment (vladopajic) [reply pkg/storageincentives/proof.go]: fixed

File: pkg/storageincentives/proof.go [hunk 7]
```diff
@@ -5,11 +5,234 @@
...
+
+	hasher, trHasher := bmtpool.Get(), bmt.NewTrHasher(anchor1)
+	defer bmtpool.Put(hasher)
+
+	anchor2Big := new(big.Int).SetBytes(anchor2)
+	segmentIndex := int(new(big.Int).Mod(anchor2Big, big.NewInt(128)).Int64())
+
+	wcp := Trio[int]{} // wcp := witness chunk position
+	wcp.A = int(new(big.Int).Mod(anchor2Big, big.NewInt(storer.SampleSize-1)).Int64())
+	wcp.B = int(new(big.Int).Mod(anchor2Big, big.NewInt(storer.SampleSize-2)).Int64())
```
[the remark this answers is not in the document]
Comment (vladopajic) [reply pkg/storageincentives/proof.go:61]: fixed

File: pkg/soc/soc.go (hunk 4 above)
Comment (vladopajic) [reply pkg/soc/soc.go:89]: Why not public fields? I didn't want to introduce big refactoring change in this PR, so I just exported existing methods. I will look in this, and potentially addressing this in separate PR.

File: pkg/storageincentives/redistribution/inclusion_proof.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (vladopajic) [reply pkg/storageincentives/redistribution/inclusion_proof.go:63]: Those tests are never reaching to the real logic on smart contract side, therefore values could be mocked. The only thing that is tested there is integration with smart contract (creating arguments, passing values and checking if those conform to contract ABI).

File: pkg/storageincentives/proof.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (aloknerurkar) [reply pkg/storageincentives/proof.go:26]: @vladopajic This hackmd link could potentially go away. I think we should add description here.

File: pkg/storer/reserve.go [hunk 8]
```diff
@@ -7,7 +7,6 @@ package storer
 import (
```
Comment (aloknerurkar) [inline pkg/storer/reserve.go:7]: I would prefer if we have a separate file for the sampler code. I think this reserve file is already pretty big. This is more inline with this pkg.

**Review by aloknerurkar — APPROVED** [review]
Apart from the minor comments LGTM!

File: pkg/storer/reserve.go (hunk 8 above)
Comment (vladopajic) [reply pkg/storer/reserve.go:7]: fixed

File: pkg/storageincentives/agent.go [hunk 9]
```diff
@@ -353,7 +361,17 @@ func (a *Agent) handleClaim(ctx context.Context, round uint64) error {
 		a.logger.Info("could not set balance", "err", err)
 	}
 
-	txHash, err := a.contract.Claim(ctx)
+	sampleData, exists := a.state.SampleData(round - 1)
+	if !exists {
+		return fmt.Errorf("sample not found")
+	}
+
+	proofs, err := makeInclusionProofs(sampleData.ReserveSampleItems, sampleData.Anchor1, sampleData.Anchor2)
```
[the remark this answers is not in the document]
Comment (nugaon) [reply pkg/storageincentives/agent.go:369]: the claim period is long and doing the computation of the proofs is fast compare to the witness filtering.
Nevertheless, the anchor should be available for the witness proofs after the first reveal on the SC API.

File: pkg/storageincentives/types/types.go [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+func NewTrio[T any](a, b, c T) Trio[T] {
+	return Trio[T]{
+		A: a,
+		B: b,
+		C: c,
+	}
+}
+
+// ToHexString returns hex encoded string from supplied slice.
+func ToHexString(data []byte) string {
```
[the remark this answers is not in the document]
Comment (nugaon) [reply pkg/storageincentives/types/types.go:25]: there should be some transformation of the sister segments to satisfy the smart contract ABI. 
either this or the Bee inclusion proofing should be refactored.

**Comment by istae** [comment]
closing as this is taken care of in #4373

**Pull request closed without merging**